File: python/tvm/meta_schedule/workload_registry.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Workload Registry"""

from typing import Any, Tuple

from tvm._ffi import register_object
from tvm.ir import IRModule
from tvm.runtime import Object

from . import _ffi_api


@register_object("meta_schedule.WorkloadToken")
class WorkloadToken(Object):
    """The class of workload tokens.

    Parameters
    ----------
    mod : tvm.ir.IRModule
        The IRModule of the workload.
    shash : int
        The structural hash of the workload.
    """

    mod: IRModule
    shash: int

    def __init__(self, mod: IRModule, shash: int, token_id: int) -> None:
        """Constructor."""
        self.__init_handle_by_constructor__(
            _ffi_api.WorkloadToken,  # type: ignore # pylint: disable=no-member
            mod,
            shash,
            token_id,
        )

    @staticmethod
    def from_json(json_obj: Any, token_id: int) -> "WorkloadToken":
        """Create a workload token from json object.

        Parameters
        ----------
        json_obj : Any
            The json object.
        token_id : int
            The token id.

        Returns
        -------
        workload_token : WorkloadToken
            The workload token.
        """
        return _ffi_api.WorkloadTokenFromJSON(json_obj, token_id)  # type: ignore # pylint: disable=no-member

    def as_json(self) -> Tuple[str, str]:
        """Export the workload token to a JSON string.

        Returns
        -------
        json_str_tuple : Tuple[str, str]
            An array containing the structural hash and the base64 json string.
        """
        shash_str, mod = _ffi_api.WorkloadTokenAsJSON(self)  # type: ignore # pylint: disable=no-member
        return (shash_str, mod)


@register_object("meta_schedule.WorkloadRegistry")
class WorkloadRegistry(Object):
    """The class for workload registry.

    Parameters
    ----------
    path : str
        The path of the workload registry.
    """

    path: str

    def __init__(self, path: str, allow_missing: bool) -> None:
        """Constructor."""
        self.__init_handle_by_constructor__(
            _ffi_api.WorkloadRegistry,  # type: ignore # pylint: disable=no-member
            path,
            allow_missing,
        )

    def lookup_or_add(self, mod: IRModule) -> WorkloadToken:
        """Lookup or add a workload.

        Parameters
        ----------
        mod : tvm.ir.IRModule
            The IRModule of the workload.

        Returns
        -------
        workload_token : WorkloadToken
            The workload token of the IRModule.
        """
        return _ffi_api.WorkloadRegistryLookupOrAdd(self, mod)  # type: ignore # pylint: disable=no-member

    def __len__(self) -> int:
        """Get size of the workload registry.

        Returns
        -------
        size : int
            The size of the workload registry.
        """
        return _ffi_api.WorkloadRegistrySize(self)  # type: ignore # pylint: disable=no-member

    def __getitem__(self, index: int) -> WorkloadToken:
        """Get a workload token by index.

        Parameters
        ----------
        index : int
            The index of the workload token.

        Returns
        -------
        workload_token : WorkloadToken
            The workload token with given index.
        """
        return _ffi_api.WorkloadRegistryAt(self, index)  # type: ignore # pylint: disable=no-member

File: src/meta_schedule/workload_registry.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./workload_registry.h"

#include <dmlc/memory_io.h>
#include <tvm/node/serialization.h>

#include "../support/base64.h"
#include "./utils.h"

namespace tvm {
namespace meta_schedule {

WorkloadToken::WorkloadToken(IRModule mod, String shash, int64_t token_id) {
  ObjectPtr<WorkloadTokenNode> n = runtime::make_object<WorkloadTokenNode>();
  n->mod = mod;
  n->shash = shash;
  n->token_id_ = token_id;
  data_ = std::move(n);
}

ObjectRef WorkloadTokenNode::AsJSON() const {
  // Convert `this->mod` to JSON
  std::string json_mod = tvm::SaveJSON(this->mod);
  // Dump the JSON string to base64
  std::string b64_mod = Base64Encode(json_mod);
  // Output
  return Array<ObjectRef>{this->shash, String(b64_mod)};
}

WorkloadToken WorkloadToken::FromJSON(const ObjectRef& json_obj, int64_t token_id) {
  IRModule mod{nullptr};
  String shash{ObjectPtr<runtime::StringObj>(nullptr)};
  try {
    const ArrayNode* json_array = json_obj.as<ArrayNode>();
    CHECK(json_array && json_array->size() == 2);
    // Load json[0] => shash
    shash = Downcast<String>(json_array->at(0));
    // Load json[1] => mod
    {
      String b64_mod = Downcast<String>(json_array->at(1));
      std::string json_mod = Base64Decode(b64_mod);
      mod = Downcast<IRModule>(LoadJSON(json_mod));
    }
    // Verify SHash(mod) == shash
    String recalc_shash = GetSHash(mod);
    CHECK_EQ(recalc_shash, shash) << "ValueError: Structural hash changed. Given: " << shash
                                  << "; Recalculated: " << recalc_shash;
  } catch (const std::runtime_error& e) {  // includes tvm::Error and dmlc::Error
    LOG(FATAL) << "ValueError: Unable to parse the JSON object: " << json_obj
               << "\nThe error is: " << e.what();
  }
  return WorkloadToken(mod, shash, token_id);
}

WorkloadToken WorkloadRegistryNode::LookupOrAdd(const IRModule& mod) {
  // Insert `mod` into the lookup table if it doesn't exist
  decltype(this->mod2token_id_)::iterator it;
  bool inserted = false;
  std::tie(it, inserted) = this->mod2token_id_.insert({mod, -1});
  int64_t& token_id = it->second;
  // Case 1. Insertion fails: it already exists in the lookup table
  if (inserted == false) {
    return this->workload_tokens_.at(token_id);
  }
  // Case 2. Insertion succeeds: add to the lookup table
  token_id = this->workload_tokens_.size();
  WorkloadToken token(mod, GetSHash(mod), token_id);
  this->workload_tokens_.push_back(token);
  JSONFileAppendLine(this->path, JSONObj2Str(token->AsJSON()));
  return token;
}

int64_t WorkloadRegistryNode::Size() const { return this->workload_tokens_.size(); }

WorkloadToken WorkloadRegistryNode::At(int64_t token_id) const {
  return this->workload_tokens_.at(token_id);
}

WorkloadRegistry::WorkloadRegistry(String path, bool allow_missing) {
  ObjectPtr<WorkloadRegistryNode> n = runtime::make_object<WorkloadRegistryNode>();
  // Set `path`
  n->path = path;
  // Set `workload_tokens_`, `mod2token_id_`
  Array<ObjectRef> json_objs = JSONStr2Obj(JSONFileReadLines(path, allow_missing));
  n->workload_tokens_.reserve(json_objs.size());
  n->mod2token_id_.reserve(json_objs.size());
  for (const ObjectRef& json_obj : json_objs) {
    int64_t token_id = n->workload_tokens_.size();
    WorkloadToken token = WorkloadToken::FromJSON(json_obj, token_id);
    n->mod2token_id_.emplace(token->mod, token_id);
    n->workload_tokens_.push_back(token);
  }
  data_ = std::move(n);
}

TVM_REGISTER_NODE_TYPE(WorkloadTokenNode);
TVM_REGISTER_NODE_TYPE(WorkloadRegistryNode);
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadToken")
    .set_body_typed([](IRModule mod, String shash, int64_t token_id) {
      return WorkloadToken(mod, shash, token_id);
    });
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadTokenAsJSON")
    .set_body_method<WorkloadToken>(&WorkloadTokenNode::AsJSON);
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadTokenFromJSON").set_body_typed(&WorkloadToken::FromJSON);
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadRegistry")
    .set_body_typed([](String path, bool allow_missing) {
      return WorkloadRegistry(path, allow_missing);
    });
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadRegistryLookupOrAdd")
    .set_body_method<WorkloadRegistry>(&WorkloadRegistryNode::LookupOrAdd);
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadRegistrySize")
    .set_body_method<WorkloadRegistry>(&WorkloadRegistryNode::Size);
TVM_REGISTER_GLOBAL("meta_schedule.WorkloadRegistryAt")
    .set_body_method<WorkloadRegistry>(&WorkloadRegistryNode::At);

}  // namespace meta_schedule
}  // namespace tvm
